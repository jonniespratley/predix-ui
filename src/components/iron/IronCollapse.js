import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

/**
 * IronCollapse component implementation.
 */
class IronCollapse extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      noAnimation: props.noAnimation || false,
      horizontal: props.horizontal || false,
      opened: props.opened || false,
      style: null
    };
    this._handleRef = this._handleRef.bind(this);
    this._transitionEndBound = this._transitionEnd.bind(this);
  }

  get dimension() {
    return this.state.horizontal
      ? 'width'
      : 'height';
  }

  /* eslint-ignore-start */
  get _isDisplayed() {
    const rect = this.base.getBoundingClientRect();
    Object.keys(rect).forEach((key) => {
      if (rect[key] !== 0) {
        return true;
      }
      return false;
    });
    return false;
  }
  /* eslint-ignore-stop */
  _calcSize() {
    return `${this.base.getBoundingClientRect()[this.dimension]}px`;
  }

  toggleClass(name, bool) {
    if (bool) {
      this.base.classList.add(name);
    } else {
      this.base.classList.remove(name);
    }
  }

  _transitionEnd() {
    if (this.state.opened) {
      this.base.style[this.dimension] = 'auto';
    }
    this._updateTransition(false);
    this.toggleClass('iron-collapse-closed', !this.state.opened);
  }

  _updateTransition(enabled) {
    this.base.style.transitionDuration = (enabled && !this.state.noAnimation)
      ? ''
      : '0s';
  }

  focus() {
    this.base.focus();
  }

  updateSize(size, animated) {
    let newSize = size;
    if (this.base.style[this.dimension] === size) {
      return;
    }
    this._updateTransition(false);
    if (animated && !this.state.noAnimation && this._isDisplayed) {
      const startSize = this._calcSize();

      // For `auto` we must calculate what is the final size for the animation.
      // After the transition is done, _transitionEnd will set the size back to `auto`.
      if (size === 'auto') {
        this.base.style[this.dimension] = size;
        newSize = this._calcSize();
      }
      this.base.style[this.dimension] = startSize;
      this.base.style.height = this.base.offsetHeight;
      this._updateTransition(true);
    }
    this.base.style[this.dimension] = newSize;
  }

  toggle() {
    if (this.state.opened) {
      this.hide();
    } else {
      this.show();
    }
  }

  /** Show content */
  show() {
    this.setState({ opened: true });
  }

  /** Hide content */
  hide() {
    this.setState({ opened: false });
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.opened !== this.state.opened) {
      this.setState({ opened: nextProps.opened });
    }
  }

  componentDidUpdate() {
    this.updateSize(this.state.opened
      ? 'auto'
      : '0px', true);
    if (this.state.opened) {
      this.focus();
    }
    if (this.state.noAnimation) {
      this._transitionEnd();
    }
    if (this.props.onChange) {
      this.props.onChange(this.state);
    }
  }

  componentDidMount() {
    if (this.base) {
      this.base.addEventListener('transitionend', this._transitionEndBound);
      this._transitionEnd();
    }
  }

  componentWillUnmount() {
    if (this.base) {
      this.base.removeEventListener('transitionend', this._transitionEndBound);
    }
  }

  _handleRef(el) {
    this.base = el;
  }

  render() {
    const { opened } = this.state;
    const { children } = this.props;
    const baseClassnames = classnames('iron-collapse');

    return (
      <div
        className={baseClassnames}
        ref={this._handleRef}
        aria-hidden={!opened}
        aria-expanded={opened}
      >
        {children}
        <style>{`
        .iron-collapse {
          display: block;
          -webkit-transition-duration: 300ms;
          -moz-transition-duration: 300ms;
          transition-duration: 300ms;
          overflow: visible;
        }
        .iron-collapse:not(.iron-collapse-opened) {
          overflow: hidden;
        }
      `}</style>
      </div>
    );
  }
}


IronCollapse.defaultProps = {
  horizontal: false,
  onChange: null,
  noAnimation: null,
  opened: false,
  children: null
};

IronCollapse.propTypes = {
  onChange: PropTypes.func,
  horizontal: PropTypes.bool,
  noAnimation: PropTypes.bool,
  opened: PropTypes.bool,
  children: PropTypes.node
};

export default IronCollapse;
