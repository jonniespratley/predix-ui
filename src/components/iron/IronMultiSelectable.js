
/* eslint-disable */
import IronSelection from './IronSelection';

export default class IronMultiSelectable extends IronSelection {
  constructor(props) {
    super(props);
    this.multi = props.multi || false;
    this.selectedValues = [];
    this.selectedItems = [];
  }
  push(prop, value) {
    this[prop].push(value);
  }
  splice(prop, index, len) {
    this[prop].splice(index, len);
  }

  /**
     * Selects the given value. If the `multi` property is true, then the selected state of the
     * `value` will be toggled; otherwise the `value` will be selected.
     *
     * @method select
     * @param {string|number} value the value to select.
     */
  select(value) {
    if (this.multi) {
      this._toggleSelected(value);
    } else {
      this.selected = value;
    }
  }
  _updateSelected() {
    if (this.multi) {
      this._selectMulti(this.selectedValues);
    } else {
      this._selectSelected(this.selected);
    }
  }
  _toggleSelected(value) {
    const i = this.selectedValues.indexOf(value);
    const unselected = i < 0;
    if (unselected) {
      this.push('selectedValues', value);
    } else {
      this.splice('selectedValues', i, 1);
    }
  }
  _valuesToItems(values) {
    return values == null
      ? null
      : values.map(function (value) {
        return this._valueToItem(value);
      }, this);
  }
  _selectMulti(values = []) {
    const selectedItems = (this._valuesToItems(values) || []).filter(function (item) {
      return item !== null && item !== undefined;
    });
    // clear all but the current selected items
    this._selection.clear(selectedItems);


    for (let i = 0; i < selectedItems.length; i++) {
      this._selection.setItemSelected(selectedItems[i], true);
    }
    // Check for items, since this array is populated only when attached
    if (this.fallbackSelection && !this._selection.get().length) {
      const fallback = this._valueToItem(this.fallbackSelection);
      if (fallback) {
        this.select(this.fallbackSelection);
      }
    }
  }
}
