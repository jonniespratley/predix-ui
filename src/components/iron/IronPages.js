import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

class IronPages extends React.Component {
  _renderChildren() {
    const {
      selected,
      selectedClassName
    } = this.props;
    const len = React.Children.count(this.props.children);
    let child = null;
    if (selected > len - 1) {
      [child] = this.props.children;
    } else {
      child = this.props.children[selected];
    }
    return (<div className={classnames(selectedClassName)}>{child}</div>);
  }

  render() {
    const {
      className
    } = this.props;

    return (
      <div className={classnames(className)}>{this._renderChildren()}</div>
    );
  }
}

IronPages.defaultProps = {
  // propForSelect: null,
  selected: 0,
  children: null,
  // selectedItem: null,
  selectedClassName: 'iron-selected',
  className: 'iron-selector'
};

IronPages.propTypes = {
  children: PropTypes.node,
  // propForSelect: PropTypes.string,
  selected: PropTypes.number,
  // selectedItem: PropTypes.string,
  selectedClassName: PropTypes.string,
  className: PropTypes.string
};

IronPages.displayName = 'IronPages';

export default IronPages;
