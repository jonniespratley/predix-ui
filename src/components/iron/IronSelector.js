import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

class IronSelector extends React.Component {
  constructor(props) {
    super(props);
    this.displayName = 'IronSelector';
    this.state = {
      selected: props.selected || 0,
      selectedClassName: props.selectedClassName || 'iron-selected',
      propForSelected: props.propForSelect || null
    };
    this._items = [];
    this._keys = [];
  }

  _renderChildren() {
    const { selected, children } = this.props;
    const child = children[selected];
    if (child) {
      return <div>{child}</div>;
    }
    return null;
  }

  componentDidUpdate() {
    if (this.props.onChange) {
      this.props.onChange(this.state);
    }
  }

  _renderTitles() {
    this._keys = [];
    function labels(child, index) {
      let propForSelect = index;
      this._items.push(child);
      if (this.props.propForSelect) {
        propForSelect = child.props[this.props.propForSelect];
        this._keys.push(propForSelect);
      } else {
        this._keys.push(index);
      }
      const { selectedClassName } = this.props;
      const selected = (this.props.selected === this._getIndexForValue(propForSelect));
      const baseClasses = classnames({ [`${selectedClassName}`]: selected });
      return (
        <li key={index} className={baseClasses}>
          {child}
        </li>
      );
    }
    const nodes = this.props.children.map(labels.bind(this));
    return nodes;
  }

  _getIndexForValue(val) {
    return this._keys.indexOf(val);
  }

  render() {
    const {
      className,
      style
    } = this.props;

    const baseClassnames = classnames(
      'iron-selector',
      className
    );

    return (
      <ul className={baseClassnames} style={style}>
        { this._renderTitles() }
      </ul>
    );
  }
}

IronSelector.defaultProps = {
  onChange: null,
  style: {
    listStyleType: 'none',
    margin: 0,
    padding: 0
  },
  propForSelect: null,
  selected: 0,
  children: null,
  // selectedItem: null,
  selectedClassName: 'iron-selected',
  className: 'iron-selector'
};

IronSelector.propTypes = {
  onChange: PropTypes.func,
  style: null,
  children: PropTypes.node,
  propForSelect: PropTypes.string,
  selected: PropTypes.number,
  // selectedItem: PropTypes.string,
  selectedClassName: PropTypes.string,
  className: PropTypes.string
};

export default IronSelector;
