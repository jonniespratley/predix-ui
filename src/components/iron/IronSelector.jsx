import React from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

class IronSelector extends React.Component {
  constructor(props) {
    super(props);
    this.displayName = 'IronSelector';
    this.state = {
      selected: props.selected,
      selectedClassName: props.selectedClassName,
      propForSelected: props.propForSelect
    };
    this._items = [];
    this._keys = [];
  }

  _renderChildren() {
    const { selected, children } = this.props;
    const child = children[selected];
    if (child) {
      return <div>{child}</div>;
    }
    return null;
  }

  componentDidUpdate() {
    const { onChange } = this.props;
    if (onChange) {
      onChange(this.state);
    }
  }

  _renderTitles = (children) => {
    this._keys = [];
    this._items = [];
    const {
      propForSelect,
      selectedClassName,
      selected
    } = this.props;

    function labels(child, index) {
      const propForSelectChild = propForSelect || index;

      this._items.push(child);
      if (propForSelectChild) {
        this._keys.push(propForSelectChild);
      } else {
        this._keys.push(index);
      }


      const selectedItem = (selected === this._getIndexForValue(propForSelectChild));
      const baseClasses = classnames({ [`${selectedClassName}`]: selectedItem });

      const propForChild = {
        [`${propForSelect}`]: child.props[`${propForSelect}`]
      };
      return (
        <li key={index} className={baseClasses} {...propForChild}>
          {child}
        </li>
      );
    }

    return children.map(labels.bind(this));
  }

  _getIndexForValue(val) {
    return this._keys.indexOf(val);
  }

  render() {
    const {
      className,
      style,
      children
    } = this.props;

    const baseClassnames = classnames(className);

    return (
      <ul className={baseClassnames} style={style}>
        { this._renderTitles(children) }
      </ul>
    );
  }
}

IronSelector.defaultProps = {
  onChange: null,
  style: {
    listStyleType: 'none',
    margin: 0,
    padding: 0
  },
  propForSelect: null,
  selected: 0,
  children: null,
  // selectedItem: null,
  selectedClassName: 'iron-selected',
  className: 'iron-selector'
};

IronSelector.propTypes = {
  onChange: PropTypes.func,
  style: null,
  children: PropTypes.node,
  propForSelect: PropTypes.string,
  selected: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  // selectedItem: PropTypes.string,
  selectedClassName: PropTypes.string,
  className: PropTypes.string
};

export default IronSelector;
