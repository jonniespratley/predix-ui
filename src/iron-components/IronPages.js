import React from 'react';
import classnames from 'classnames';



class IronPages extends React.Component {
	constructor(props){
		super(props);
		this.state = {
			selected: props.selected || 0,
			selectedItem: null,
			attrForSelected: props.attrForSelected || null
		};
	}
	_renderChildren(){
		let selected = this.props.selected;
		let child = null;
		let len = this.props.children.length;
		if(selected > len - 1){
			child = this.props.children[0];
		} else {
			child = this.props.children[selected];
		}


    const selectedClassName = classnames(this.props.selectedClassName);

		return (<div className={selectedClassName}>{child}</div>);
	}
	render(){

    const {
      children,
      className = 'iron-selector',
      selectedClassName = 'iron-selected',
      style = null
    } = this.props;

    const { selected, selectedItem } = this.state;

    const baseClassnames = classnames(className);

		return (<div className={baseClassnames}>{this._renderChildren()}</div>)
	}
}
IronPages.defaultProps = {
  selectedClassName: 'iron-selected'
};

export default IronPages;
