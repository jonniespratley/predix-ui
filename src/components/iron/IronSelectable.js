/* eslint-disable */
import IronMultiSelectable from './IronMultiSelectable';
/**
 * IronSelectable - https://github.com/PolymerElements/iron-selector/blob/master/iron-selectable.html
   <iron-selector selected="0">
    <div>Item 1</div>
    <div>Item 2</div>
    <div>Item 3</div>
  </iron-selector>
 */
export default class IronSelectable extends IronMultiSelectable {
  constructor(props = {}) {
    super(props);
    this.items = props.children || props.items || [];
    this.attrForSelected = props.attrForSelected || null;
    this.selectedClassName = props.selectedClassName || 'iron-selected';
    this.selectedItem = props.selectedItem || null;
    this.selected = props.selected || null;
    this.selectedAttribute = props.selectedAttribute || null;
  }
  indexOf(item) {
    return this.items.indexOf(item);
  }
  select(value) {
    this.selected = value;
    this.selectedItem = value;
  }
  selectPrevious() {
    const { length } = this.items;
    const index = (Number(this._valueToIndex(this.selected)) - 1 + length) % length;
    this.select(this._indexToValue(index));
  }
  selectNext() {
    const index =
			(Number(this._valueToIndex(this.selected)) + 1) % this.items.length;
    this.select(this._indexToValue(index));
  }
  selectIndex(index) {
    this.select(this._indexToValue(index));
  }
  _valueToIndex(value) {
    if (this.attrForSelected) {
      for (var i = 0, item; (item = this.items[i]); i++) {
        if (this._valueForItem(item) == value) {
          return i;
        }
      }
    } else {
      return Number(value);
    }
  }
  _indexToValue(index) {
    if (this.attrForSelected) {
      const item = this.items[index];
      if (item) {
        return this._valueForItem(item);
      }
    } else {
      return index;
    }
  }
  _valueForItem(item) {
    if (!item) {
      return null;
    }
    const propValue = item.props[this.attrForSelected];
    return propValue != undefined ? propValue : item;
  }
  _valueForIndex(index) {
    if (!index) {
      return null;
    }
    const propValue = this.items[this.attrForSelected];
    return propValue != undefined ? propValue : this.items[index];
  }
  _selectionChange() {
    this._setSelectedItem(this._selection.get());
  }
}
