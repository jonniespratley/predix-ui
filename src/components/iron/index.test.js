import {expect} from 'chai';
import React from 'react';
import {shallow, mount} from 'enzyme';
import IronComponents from './';

import IronCollapse from './IronCollapse';
import IronSelector from './IronSelector';
import IronPages from './IronPages';

import IronSelectable from './IronSelectable';
import IronMultiSelectable from './IronMultiSelectable';

function createMockItem(i) {
	return {
		props: {
			id: `page-${i}`,
			name: "Item " + i
		}
	};
}

describe('iron', () => {

  describe('IronSelectable', () => {
    test('should set selectedItem', () => {
      const instance = new IronSelectable({
        multi: false,
        selected: 1,
        items: [createMockItem(1), createMockItem(2), createMockItem(3), createMockItem(4), createMockItem(5)]
      });
      instance.selectNext();
      expect(instance.selected).to.equal(2);
      instance.selectNext();
      expect(instance.selected).to.equal(3);
      instance.selectPrevious();
      expect(instance.selected).to.equal(2);
    });

    test('should select item in attrForSelected', () => {
      const instance = new IronSelectable({
        multi: false,
        attrForSelected: 'id',
        selected: 'page-1',
        items: [createMockItem(1), createMockItem(2), createMockItem(3), {props: {id: 1}}]
      });
      instance.selectNext();
      expect(instance.selectedItem).to.equal('page-2');
      instance.selectNext();
      expect(instance.selectedItem).to.equal('page-3');
      instance.selectPrevious();
      instance.selectPrevious();
      expect(instance.selectedItem).to.equal('page-1');
      //instance.selectNext();
      //expect(instance.selectedItem).to.equal('page-3');
    });
  });

  xdescribe('iron-collapse', () => {
    test('should render', () => {
      const wrapper = shallow(<IronCollapse/>);
      console.log(wrapper.debug());
      expect(wrapper.find('.iron-collapse')).to.have.length(1);
    });
  });

  describe('iron-pages', () => {
    test('should render selected page', () => {
      const wrapper = shallow(
        <IronPages selected={0}>
          <div>One</div>
          <div>Two</div>
          <div>Three</div>
        </IronPages>
      );
      expect(wrapper.find('.iron-selected')).to.have.length(1);
    });
  });

  describe('iron-selector', () => {
    test('should render selected item', () => {
      const wrapper = shallow(
        <IronSelector selected={1}>
          <div>1</div>
          <div>2</div>
          <div>3</div>
        </IronSelector>
      );
      console.log(wrapper.debug());
      expect(wrapper.find('.iron-selected')).to.have.length(1);
    });
  });

  //expect(wrapper.find('.label')).to.have.length(1);
  //expect(wrapper.find('.delta')).to.have.length(1);
  //expect(wrapper.find('.alpha')).to.have.length(1);
  //expect(wrapper.contains(<div className='label'/>)).to.equal(true);
});
