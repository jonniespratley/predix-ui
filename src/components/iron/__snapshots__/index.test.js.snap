// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Iron Components IronPages should render selected page 1`] = `
<div
  className="iron-selector"
>
  <div
    className="iron-selected"
  >
    <div>
      One
    </div>
  </div>
</div>
`;
